include/Dialects/LinalgTransform/TransformOpInterface.h

```diff
@@ -51,27 +51,106 @@ class TransformState {
   protected:
-    // FIXME: this provides direct write access to the state internals, replace
-    // with more transactional methods.
-    TransformOpMapping &getMapping(TransformState &state) {
-      return state.operations;
+    Extension(TransformState &state) : state(state) {}
+
+    /// Read-only access to the mapping between transform IR values and payload
+    /// IR operations contained in the state.
+    const TransformOpMapping &getMapping() const { return state.operations; }
+
+    /// Notifies the extension that payload IR operations were associated with
+    /// the given transform IR handle. Concrete extensions that are willing to
+    /// be notified should override this method.
+    virtual void notifySetPayload(Value handle,
+                                  ArrayRef<Operation *> operations) {}
+    /// Notifies the extension that the association between a transform IR
+    /// handle and a list of payload IR operations is about to be removed.
+    /// Concrete extensions that are willing to be notified should override this
+    /// method.
+    virtual void notifyRemovePayload(Value handle,
+                                     ArrayRef<Operation *> operations) {}
+
+    /// Notifies the extension that the ops associated with the transform IR
+    /// handle changed. Concrete extensions that are willing not be notified
+    /// should override this method.
+    virtual void notifyUpdatePayload(Value handle, ArrayRef<Operation *> oldOps,
+                                     ArrayRef<Operation *> newOps) {}
+
+    /// Sets the payload IR ops associated with the given transform IR value.
+    /// Fails if this would result in multiple transform IR values with uses
+    /// corresponding to the same payload IR ops. This extension will NOT
+    /// be notified about this event.
+    LogicalResult setPayloadOps(Value handle,
+                                ArrayRef<Operation *> operations) {
+      propagatingSetPayload = true;
+      LogicalResult result = state.setPayloadOps(handle, operations);
+      propagatingSetPayload = false;
+      return result;
+    }
+
+    /// Forgets the payload IR ops associated with the given transform IR value.
+    /// This extension will NOT be notified about this event.
+    void removePayloadOps(Value handle) {
+      propagatingRemovePayload = true;
+      state.removePayloadOps(handle);
+      propagatingRemovePayload = false;
     }
+
+    /// Updates the payload IR ops associated with the given transform IR value.
+    /// The callback function is called once per associated operation and is
+    /// expected to return the modified operation or nullptr. In the latter
+    /// case, the corresponding operation is no longer associated with the
+    /// transform IR value. This extension will NOT be notified about it.
+    void updatePayloadOps(Value handle,
+                          function_ref<Operation *(Operation *)> callback) {
+      return state.updatePayloadOps(handle, callback);
```

Comment:
I would expect propagatingUpdatePayload needs to be set?

Reply:
Nice catch, thanks!